Feature: Automatic UI and font zoom levels.

michicc opened the pull request:
## Motivation / Problem

On HiDPI screens, the default GUI and font zoom might be too small to read.


## Description

An additional (and default) "auto" value for the zoom settings is added, that will get the current DPI value/scaling from the video driver and choose a zoom value accordingly.

For Win32 and OSX, the set thresholds are 2X zoom at a DPI scale > 150% and 4X zoom at a scale > 300%. Current OSX version will only report 100% or 200% for the scale factor, but the implementation still checks for 300% to allow for future OSX updates.


## Limitations

I've marked this as a draft, as the OSX zoom level suggestion as implemented in this PR right now only makes sense together with #8519. It is it's own commit, so it could easily be stripped off for now.

It is also missing a SDL implementation, at least if HiDPI is a thing with SDL.


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* ~The bug fix is important enough to be backported? (label: 'backport requested')~
* ~This PR affects the save game format? (label 'savegame upgrade')~
* ~This PR affects the GS/AI API? (label 'needs review: Script API')~
* ~This PR affects the NewGRF API? (label 'needs review: NewGRF')~

michicc committed:
Codechange: Make the settings for min/max zoom early load settings.

michicc committed:
Feature: Automatic UI and font zoom levels when supported by the OS.

michicc committed:
Add: [Win32] Automatic zoom level suggestion for Win32 video driver.

The zoom level suggestion is based on the DPI scaling set in Windows.
We use 150% scaling as the threshold for 2X zoom and 300% scaling
as the threshold for 4X zoom.

michicc commented:
Stripped OSX for now, will tack it on #8519 later.

frosch123 commented on src/video/win32_v.cpp:
```diff
@@ -1350,3 +1350,39 @@ Dimension VideoDriver_Win32::GetScreenSize() const
+ZoomLevel VideoDriver_Win32::GetSuggestedUIZoom()
+{
+	typedef UINT (WINAPI *PFNGETDPIFORWINDOW)(HWND hwnd);
+	typedef UINT(WINAPI *PFNGETDPIFORSYSTEM)(VOID);
```
TrueBrain stole your space.

michicc commented:
Improved Win32 code a bit.

michicc committed:
Codechange: [Win32] Increase SDK version for 32-bit target.

Use the same Windows XP target as for 64-bit. Current MSVC version will
not produce a binary that works on anything earlier anyway.

TrueBrain commented on src/video/win32_v.cpp:
```diff
@@ -1316,3 +1316,45 @@ Dimension VideoDriver_Win32::GetScreenSize() const
+		_GetDpiForMonitor = (PFNGETDPIFORMONITOR)GetProcAddress(LoadLibrary(_T("Shcore.dll")), "GetDpiForMonitor");
+	}
+
+	const UINT BASE_DPI = 96; // Default Windows DPI value.
```
Maybe either move this out of the function or closer to where it is used? There is a bit of a distance now, and it is in the middle of a function. Was a bit hard to spot.

TrueBrain commented on src/video/win32_v.cpp:
```diff
@@ -1316,3 +1316,45 @@ Dimension VideoDriver_Win32::GetScreenSize() const
+{
+	typedef UINT (WINAPI *PFNGETDPIFORWINDOW)(HWND hwnd);
+	typedef UINT (WINAPI *PFNGETDPIFORSYSTEM)(VOID);
+	typedef HRESULT (WINAPI *PFNGETDPIFORMONITOR)(HMONITOR hMonitor, int dpiType, UINT *dpiX, UINT *dpiY);
```
Is this normal for OpenTTD, to have these kind of `typedef`s inside a function?

nielsmh replied:
I think it's reasonable to do for this case at least. When you're feature-detecting for functions that may only exist in a higher API level than you're statically linking for, it makes sense to make the scope as narrow as possible.

TrueBrain commented on src/zoom_type.h:
```diff
@@ -15,6 +15,8 @@
 static uint const ZOOM_LVL_SHIFT = 2;
 static uint const ZOOM_LVL_BASE  = 1 << ZOOM_LVL_SHIFT;
 
+static const int8 ZOOM_LVL_GUI_AUTO = -1;
```
Putting this out of context of `ZOOM_LVL_MIN` is a bit awkward, as core earlier assumes that `ZOOM_LVL_GUI_AUTO` is below `ZOOM_LVL_MIN` (as any value below `ZOOM_LVL_MIN` assumes it was auto-detect). But here there is no such link or indication the values are linked like this. Can I suggest to either use `ZOOM_LVL_MIN` in the value of `ZOOM_LVL_GUI_AUTO`, or compare earlier against this value (instead of `< ZOOM_LVL_MIN`?)

TrueBrain commented on src/settings_gui.cpp:
```diff
@@ -333,8 +335,8 @@ struct GameOptionsWindow : Window {
 			case WID_GO_RESOLUTION_DROPDOWN: SetDParam(0, GetCurRes() == _resolutions.size() ? STR_GAME_OPTIONS_RESOLUTION_OTHER : SPECSTR_RESOLUTION_START + GetCurRes()); break;
-			case WID_GO_GUI_ZOOM_DROPDOWN:   SetDParam(0, _gui_zoom_dropdown[ZOOM_LVL_OUT_4X - _gui_zoom]); break;
-			case WID_GO_FONT_ZOOM_DROPDOWN:  SetDParam(0, _font_zoom_dropdown[ZOOM_LVL_OUT_4X - _font_zoom]); break;
+			case WID_GO_GUI_ZOOM_DROPDOWN:   SetDParam(0, _gui_zoom_dropdown[_gui_zoom_cfg != ZOOM_LVL_GUI_AUTO ? ZOOM_LVL_OUT_4X - _gui_zoom_cfg + 1 : 0]); break;
```
one moment you do `a == b ? 0 : c` , the next you do `a != b ? c : 0` :D Maybe we should be consistent here. Minor nitpick :)

TrueBrain commented on src/settings_gui.cpp:
```diff
@@ -538,7 +540,8 @@ struct GameOptionsWindow : Window {
 				GfxClearSpriteCache();
-				_gui_zoom = (ZoomLevel)(ZOOM_LVL_OUT_4X - index);
+				_gui_zoom_cfg = index > 0 ? ZOOM_LVL_OUT_4X - index + 1 : ZOOM_LVL_GUI_AUTO;
+				UpdateGUIZoom();
```
What the old code already didn't, and which is rather annoying from a user experience: when the option doesn't change, it still reloads everything (which hangs the game for a bit). Shouldn't we check here if the value really changed and ignore updates if it does? Possibly a bit outside the current PR, but it is one of those annoying things :P

TrueBrain commented on src/settings_gui.cpp:
```diff
@@ -547,7 +550,8 @@ struct GameOptionsWindow : Window {
 			case WID_GO_FONT_ZOOM_DROPDOWN:
 				GfxClearSpriteCache();
-				_font_zoom = (ZoomLevel)(ZOOM_LVL_OUT_4X - index);
+				_font_zoom_cfg = index > 0 ? ZOOM_LVL_OUT_4X - index + 1 : ZOOM_LVL_GUI_AUTO;
```
In general I like this PR, but the implementation feels a bit awkward, and I have been wondering why I feel that way. And I think it comes down to the way that "gui" variables and non-"gui" variables are now mixed with signed vs unsigned variables and with `< ZOOM_LVL_MIN` with `ZOOM_LVL_GUI_AUTO`.

I think you went this road because of the settings system, where you need a `min` and `max` for every setting, and didn't want to use a value like `4` for auto-detect. Which I can understand. But the code now becomes a bit difficult to follow, more so than it already does (the zoom-code is really hard to follow, because of the way it was implemented). But bumping all current zoom-levels by one is also sub-optimal, of course.

Two things that stand out: `ZOOM_LVL_GUI_AUTO` seems like a weird name to pick, as it conflicts with `gui_zoom` vs `_font_zoom`. Suggestion: `ZOOM_LVL_AUTO` or `ZOOM_LVL_CFG_AUTO`.

Next, is a bit above, where you do `if (_gui_zoom_cfg < ZOOM_LVL_MIN) {`, which implies that `ZOOM_LVL_GUI_AUTO` should be a lower value than `ZOOM_LVL_MIN`, but as I mentioned in another comment, is not enforced in the code itself. Suggestion: `if (_gui_zoom_cfg == ZOOM_LVL_GUI_AUTO) {`.

I think that these changes alone are enough to make it a bit more clear what is going on.

TrueBrain commented on src/lang/english.txt:
```diff
@@ -1004,13 +1004,15 @@ STR_GAME_OPTIONS_RESOLUTION_OTHER                               :other
 STR_GAME_OPTIONS_FONT_ZOOM                                      :{BLACK}Font size
 STR_GAME_OPTIONS_FONT_ZOOM_DROPDOWN_TOOLTIP                     :{BLACK}Select the interface font size to use
 
+STR_GAME_OPTIONS_FONT_ZOOM_DROPDOWN_AUTO                        :Auto
```
`Auto` feels weird. I think `Auto-detect` would already be better? Or `(auto-detect)`? Or `(depending on system DPI)`. I dunno .. I like `()` around these values to indicate they are not real value, but I am not hang up on that. Just only the word `Auto` doesn't cut it :)

TrueBrain reviewed (commented):
In general, I think we should do something like this.

But, there is a but: we only have this for Windows implemented now. The commit message "Feature: .." suggests it is implemented for all. On all video drivers you can see the "auto" value. Yet, on most, it doesn't really do anything.

Not sure how to deal with this. Possibly simply change the feature into making it clear it is only for a select subset of video-drivers, but .. dunno :)

michicc commented:
General comment: Per previous IRC chat, apparently SDL (at least as currently used by us) has no concept of a HiDPI window, which means nothing to scale. On OSX, we currently force OS scaling, automatic zoom would only be a thing together with #8519.

Specific comments will be dealt with later.

TrueBrain commented:
> 
> 
> General comment: Per previous IRC chat, apparently SDL (at least as currently used by us) has no concept of a HiDPI window, which means nothing to scale. On OSX, we currently force OS scaling, automatic zoom would only be a thing together with #8519.
> 

For a moment I was thinking: maybe we shouldn't show the "auto" option for SDL, for example. But I realised ... nobody is going to care if there is an "auto" option in there, that doesn't do anything for their driver. They simply will not notice. So that is just extra work for no added benefit. So yeah, forget about all that, let's just make this a generic "feature" on which we should extend.

For example, SDL does have HiDPI support just fine, but we don't use it. There is a bit of work to do for us to support it, but not much. So we might as well. There are functions like `SDL_GetDisplayDPI` which are implemented for Windows, X11, and of course mac OS. I might give a crack at this next week .. as why not!

TrueBrain commented on src/video/win32_v.cpp:
```diff
@@ -1316,3 +1316,45 @@ Dimension VideoDriver_Win32::GetScreenSize() const
+	}
+	if (cur_dpi == 0 && _GetDpiForSystem != nullptr) {
+		/* Fall back to system DPI. */
+		UINT dpi = _GetDpiForSystem();
```
This statement is not doing anything btw; should it be `cur_dpi = _GetDpiForSystem();`?

TrueBrain commented:
> 
> For example, SDL does have HiDPI support just fine, but we don't use it. There is a bit of work to do for us to support it, but not much. So we might as well. There are functions like `SDL_GetDisplayDPI` which are implemented for Windows, X11, and of course mac OS. I might give a crack at this next week .. as why not!

So I was thinking this would be simple, SDL is giving us a function, so how hard could it be? Well, for Windows, it is fine. But we of course mostly target Linux with SDL ... euh .. yeah ...

Someone worked out nicely how SDL behaves here, which you can read here:
https://github.com/mosra/magnum/commit/ae31c3cd82ba53454b8ab49d3f9d8ca385560d4b

And if we look in the SDL source, we found on X11 `SDL_GetDisplayDPI` gets its values from here:
https://github.com/libsdl-org/SDL/blob/main/src/video/x11/SDL_x11modes.c#L473
https://github.com/libsdl-org/SDL/blob/main/src/video/x11/SDL_x11modes.c#L756

Long story short: `SDL_GetDisplayDPI` will never return a useful result on Linux. A good solution is to use `Xft.dpi`, which is exactly what we would like (and the same as in the Win32 driver), but it requires some magic with `dl`:
https://github.com/mosra/magnum/commit/ae31c3cd82ba53454b8ab49d3f9d8ca385560d4b#diff-953f069b28302b3aa5079d9fe56f4e646c93e52a09891554d4150f3dd2ebd0c4R64

This would be similar to how Win32 driver now solves it, so that might be okay, but do we want to go that route, I wonder?

michicc committed:
Codechange: Re-init GUI zoom only when really changed.

michicc committed:
Add: [OSX] Automatic zoom level suggestion for Cocoa video driver.

TrueBrain reviewed (approved):
I will see what I can do for SDL this week; so we can either hold off on this PR till that is done, or just PR it later. I think the latter works just fine.
